File: ext/rubypyapi/py_object.c
#include "py_object.h"

#include "ptor.h"
#include "rtop.h"

RUBY_EXTERN VALUE mRubyPyApi;

VALUE cRubyPyObject;

static VALUE PyStructAlloc(VALUE);
static void PyStructFree(PyStruct*);


//Create a new PyStruct
static
VALUE PyStructAlloc(VALUE klass)
{
	PyStruct* self = ALLOC(PyStruct);
	self->pObject = NULL;
	
	return Data_Wrap_Struct(klass, 0, PyStructFree, self);
}


//Delete a PyStruct
static
void PyStructFree(PyStruct* self)
{
	if(Py_IsInitialized() && self->pObject)
	{
		//Make sure we decrement the object count on our wrapped
		//object before we free the ruby wrapper
		Py_XDECREF(self->pObject);
	}
	free(self);
}

static
VALUE rpRubify(VALUE self) {
    VALUE rbObject;
    PyObject* pObject;
    PyStruct* pyStruct;
    
    Data_Get_Struct(self, PyStruct, pyStruct);
    
    pObject = pyStruct->pObject;
    
    rbObject = ptorObject(pObject);
    
    return rbObject;
}

static
VALUE PyStructInit(VALUE self, VALUE rbObject) {
	PyObject* pObject;
	PyStruct* cSelf;
	
	pObject = rtopObject(rbObject, 0);
	
	Data_Get_Struct(self, PyStruct, cSelf);
	
	//Py_XDECREF(cSelf->pObject);
	cSelf->pObject = pObject;

	return self;
}

static
VALUE rpHasAttr(VALUE self, VALUE attrName) {
	PyStruct* cSelf;
	char* cName;
	VALUE has;
	
	cName = StringValueCStr(attrName);
	
	Data_Get_Struct(self, PyStruct, cSelf);
	
	if(PyObject_HasAttrString(cSelf->pObject, cName)) {
		has = Qtrue;
	} else {
		has = Qfalse;
	}
	
	return has;
	
}

static
VALUE rpGetAttr(VALUE self, VALUE attrName) {
	PyStruct* cSelf;
	PyStruct* cAttr;
	PyObject* pyAttr;
	char* cName;
	VALUE rbAttr;
	
	cName = StringValueCStr(attrName);
	
	Data_Get_Struct(self, PyStruct, cSelf);
	
	pyAttr = PyObject_GetAttrString(cSelf->pObject, cName);
	
	rbAttr = rb_obj_alloc(cRubyPyObject);
	
	Data_Get_Struct(rbAttr, PyStruct, cAttr);
	
	cAttr->pObject = pyAttr;
	
	return rbAttr;
	
}

static
VALUE rpSetAttr(VALUE self, VALUE attrName, VALUE rbPyAttr) {
	PyStruct* cSelf;
	PyStruct* cAttr;
	char* cName;
	
	cName = StringValueCStr(attrName);
	
	Data_Get_Struct(self, PyStruct, cSelf);
	Data_Get_Struct(rbPyAttr, PyStruct, cAttr);
	
	PyObject_SetAttrString(cSelf->pObject, cName, cAttr->pObject);
	
	return Qtrue;
}

static
VALUE rpCallObject(VALUE self, VALUE rbPyArgs) {
	PyStruct* cSelf;
	PyStruct* cReturn;
	PyStruct* cArgs;
	PyObject* pyReturn;
	VALUE rbReturn;
	
	Data_Get_Struct(rbPyArgs, PyStruct, cArgs);
	
	Data_Get_Struct(self, PyStruct, cSelf);
	
	pyReturn = PyObject_CallObject(cSelf->pObject, cArgs->pObject);
	
	rbReturn = rb_obj_alloc(cRubyPyObject);
	
	Data_Get_Struct(rbReturn, PyStruct, cReturn);
	
	cReturn->pObject = pyReturn;
	
	return rbReturn;

}

static
VALUE rpXDECREF(VALUE self) {
	PyStruct* cSelf;

	Data_Get_Struct(self, PyStruct, cSelf);
	
	Py_XDECREF(cSelf->pObject);
	
	cSelf->pObject = NULL;
	
	return self;
}

static
VALUE rpXINCREF(VALUE self) {
	PyStruct* cSelf;

	Data_Get_Struct(self, PyStruct, cSelf);
	
	Py_XINCREF(cSelf->pObject);
	
	return self;
}

static
VALUE rpIsNull(VALUE self) {
  PyStruct* cSelf;
  Data_Get_Struct(self, PyStruct, cSelf);

  if(cSelf->pObject)
    return Qfalse;
  else
    return Qtrue;
}

static
VALUE rpCompare(VALUE self, VALUE other) {
  PyStruct *cSelf, *cOther;
  VALUE rResult;
  int cResult;
  
  Data_Get_Struct(self, PyStruct, cSelf);
  Data_Get_Struct(other, PyStruct, cOther);

  cResult = PyObject_Compare(cSelf->pObject, cOther->pObject);

  rResult = INT2NUM(cResult);

  return rResult;
  
}

inline void Init_RubyPyObject() {
	cRubyPyObject = rb_define_class_under(mRubyPyApi,"PyObject", rb_cObject);
        rb_define_alloc_func(cRubyPyObject, PyStructAlloc);
	rb_define_method(cRubyPyObject, "initialize", &PyStructInit, 1);
        rb_define_method(cRubyPyObject, "rubify", &rpRubify, 0);
	rb_define_method(cRubyPyObject, "hasAttr", &rpHasAttr, 1);
	rb_define_method(cRubyPyObject, "getAttr", &rpGetAttr, 1);
	rb_define_method(cRubyPyObject, "setAttr", &rpSetAttr, 2);
	rb_define_method(cRubyPyObject, "callObject", &rpCallObject, 1);
	rb_define_method(cRubyPyObject, "xDecref", &rpXDECREF, 0);
	rb_define_method(cRubyPyObject, "xIncref", &rpXINCREF, 0);
	rb_define_method(cRubyPyObject, "null?", &rpIsNull, 0);
	rb_define_method(cRubyPyObject, "cmp", &rpCompare, 1);
}

File: lib/rubypython/rubypyapi/py_object.rb
require 'rubypython/rubypyapi/python'
require 'rubypython/rubypyapi/macros'
require 'rubypython/rubypyapi/ptor'
require 'rubypython/rubypyapi/rtop'
require 'ffi'

module RubyPyApi

  #This object is an opaque wrapper around the C PyObject* type used by the python
  #C API. This class <em>should not</em> be used by the end user. They should instead
  #make use of the RubyPyApi::RubyPyProxy class and its subclasses.
  class PyObject

    #This class wraps C PyObject*s so that the their Python reference count is
    #automatically decreased when the Ruby object referencing them 
    #goes out of scope.
    class AutoPyPointer < FFI::AutoPointer
      def self.release(pointer)
        Python.Py_DecRef pointer if Python.IsInitialized == 1 
      end
    end

    attr_reader :pointer

    #@param [FFI::Pointer, other] pointer objects passed in to the constructor
    #   are just assigned to the pointer attribute of the instance. All other
    #   objects are converted via {RTOP#rtopObject} before being assigned.
    def initialize(rObject)
      if rObject.kind_of? FFI::Pointer 
        @pointer = AutoPyPointer.new rObject
        xIncref if rObject.is_a? AutoPyPointer
      else
        @pointer = AutoPyPointer.new RTOP.rtopObject(rObject)
      end
    end

    #Attempts to convert the wrapped object to a native ruby type.
    #@return a ruby version of the wrapped object
    #@raise [{PTOR::UnsupportedConversion}]
    def rubify
      PTOR.ptorObject @pointer
    end

    #Tests whether the wrapped object has a given attribute
    #@param [String] the name of the attribute to look up
    #@return [Boolean]
    def hasAttr(attrName)
      Python.PyObject_HasAttrString(@pointer, attrName) == 1
    end

    #Retrieves an object from the wrapped python object
    #@param [String] the name of attribute to fetch
    #@return [{PyObject}] a Ruby wrapper around the fetched attribute
    def getAttr(attrName)
      pyAttr = Python.PyObject_GetAttrString @pointer, attrName
      self.class.new pyAttr
    end

    def setAttr(attrName, rbPyAttr)
      Python.PyObject_SetAttrString(@pointer, attrName, rbPyAttr.pointer) != -1
    end

    def callObject(rbPyArgs)
      pyReturn = Python.PyObject_CallObject(@pointer, rbPyArgs.pointer)
      self.class.new pyReturn
    end

    def xDecref
      @pointer = FFI::Pointer::NULL
    end

    def xIncref
      Python.Py_IncRef @pointer
    end

    def null?
      @pointer.null?
    end

    def cmp(other)
      Python.PyObject_Compare @pointer, other.pointer 
    end

    def functionOrMethod?
      isFunc = (Macros.PyObject_TypeCheck(@pointer, Python.PyFunction_Type.to_ptr) != 0)
      isMethod = (Macros.PyObject_TypeCheck(@pointer, Python.PyMethod_Type.to_ptr) != 0)
      isFunc or isMethod
    end

    def callable?
      Python.PyCallable_Check(@pointer) != 0
    end

    def class?
      isClassObj = (Macros.PyObject_TypeCheck(@pointer, Python.PyClass_Type.to_ptr) == 1)
      isTypeObj = (Macros.PyObject_TypeCheck(@pointer, Python.PyType_Type.to_ptr) == 1)
      isTypeObj or isClassObj
    end

    def self.makeTuple(rbObject)
      pTuple = nil

      if Macros.PyObject_TypeCheck(rbObject.pointer, Python.PyList_Type.to_ptr) != 0
        pTuple = Python.PySequence_Tuple(rbObject.pointer)
      elsif Macros.PyObject_TypeCheck(rbObject.pointer, Python.PyTuple_Type.to_ptr) != 0
        pTuple = rbObject.pointer
      else
        pTuple = Python.PyTuple_Pack(1, :pointer, rbObject.pointer)
      end

      self.new pTuple
    end

    def self.newList(*args)
      rbList = self.new Python.PyList_New(args.length)

      args.each_with_index do |el, i|
        Python.PyList_SetItem rbList.pointer, i, el.pointer
      end

      rbList
    end

    def self.convert(*args)
      args.map! do |arg|
        if arg.instance_of? RubyPyApi::PyObject
          arg
        elsif(arg.instance_of? RubyPyApi::RubyPyProxy)
          arg.pObject
        else
          RubyPyApi::PyObject.new arg
        end
      end
    end

    def self.buildArgTuple(*args)
      pList = RubyPyApi::PyObject.newList(*args)
      pTuple = RubyPyApi::PyObject.makeTuple(pList)
      pList.xDecref
      pTuple
    end

  end

end
